autoendpoint/src/routers/fcm/router.rs

```diff
@@ -70,6 +72,51 @@ impl FcmRouter {
     pub fn active(&self) -> bool {
         !self.clients.is_empty()
     }
+
+    /// Do the gauntlet check to get the router credentials.
+    /// FCM stores the values in the top hash as `token` & `app_id`, GCM stores them
+    /// in a sub-hash as `creds.auth` and `creds.senderID`.
+    /// If any of these error out, it's probably because of a corrupted key.
+    fn get_creds(&self, router_data: &HashMap<String, Value>) -> ApiResult<(String, String)> {
+        let creds = router_data.get("creds").and_then(Value::as_object);
+        // I'm sure that there's a clever way to collapse these using `.map` or `.and_then`, but
+        // this is readable to me.
+        let auth = match router_data.get("token").and_then(Value::as_str) {
+            Some(v) => v.to_owned(),
+            None => {
+                if creds.is_none() {
+                    return Err(FcmError::NoRegistrationToken.into());
+                }
+                match creds
```

Comment:
Is pulling from `creds.get("auth")` even necessary? The Python gcm.py router seems to always use `router_data.get("token")`.

Reply:
GCM stores [the server credential auth token](https://github.com/mozilla-services/autopush/blob/master/autopush/router/gcm.py#L70) as `auth`. This is different from the [registration id value](https://github.com/mozilla-services/autopush/blob/master/autopush/router/gcm.py#L110) stored in `token`

Reply:
Right, it's stored as `auth` in `router_data`, but autopush python never actually reads from it. Instead the router reads `auth` from its own copy supplied from settings at startup (as we do in rust, reading FCMClient's `credential.credential`).

Whereas here we seem to be reading it as a fallback for `router_data.get("token")` which is `registration_id`, not `auth` (plus I don't think `router_data.get("token")` should ever be returning `None`)

Also this value's called `auth` in `send_gcm`, maybe call it `token` or `registration_id` instead?.

Reply:
So, I think a BIG part of this is that the variables have terrible names. I'm going to try and fix that.

(_The following is also for my benefit..._)
What gets stored into the routing information is the `project_id` and the users `routing_token`. The `project_id` identifies which project to use, and is later matched with the `project_id` in the settings to get the corresponding `server_access_token` (which may be a value, a JSON block, or a file descriptor). FCM uses an `authenticator` function, and GCM needs to have access to the `server_credentials`, so both of those are stored into the `FcmClient` when it's initialized.